### PR TITLE
refactor(util.uri): refactor dict comprehension

### PR DESCRIPTION
**Pull request by withshubh**
This PR fixes some of the quality issues raised by DeepSource on my fork of this repository.

**Some of them are:**
- **Performance Issues: 4**
- **Bug Risks: 14**
- **Anti-Patterns: 24**
- **Security: 6**

Take a quick look at all the issues caught by DeepSource for this repository [here](https://deepsource.io/gh/withshubh/falcon/issues/?category=recommended)

# Summary of Changes

- ~~Change methods not using its bound instance to staticmethods~~
- ~~use literal syntax instead of function calls to create data structure~~
- ~~remove assert statement from non-test files~~
- ~~refactor unnecessary 'else' / 'elif' when 'if' block has a 'return' statement~~ 
- ~~replace ternary syntax with if expression~~
- ~~remove methods with unnecessary super delegation~~
- remove unnecessary generator

You can also have a look at the [configuration file](https://github.com/withshubh/falcon/blob/master/.deepsource.toml) I used for DeepSource Analysis.

If any of these issues that the DeepSource analysis found are critical for the codebase I'd suggest you to activate the analysis. You can trigger analysis on each PR once you integrate DeepSource in the repo. ⚡

Feel free to merge this PR if you wish to fix the issues.✨

### CODE REVIEW

**Comment by vytas7** [comment]
Hi, and thanks for this!

I'm, however, unsure if we want to refactor everything to `@staticmethod`. In some places, methods semantically belong to a class instance, even if they do not use self. In others, it probably does not warrant a change.

@kgriffs @CaselIT thoughts?

**Comment by CaselIT** [comment]
I'm definitely -1 on the static method thing.

**Comment by vytas7** [comment]
Another comment unrelated to the use of `@staticmethod`.

Sorry to be blunt, but refactoring `assert` statements to manually raising `AssertionError` makes no sense to me.
The `assert` statement **is** the canonical way of [writing tests with `pytest`](https://docs.pytest.org/en/latest/getting-started.html#create-your-first-test). Manually raising an `AssertionError` in a conditional branch may even interfere with `pytest`'s ability to infer useful failure messages.

File: falcon/util/uri.py
Comment (vytas7) [inline falcon/util/uri.py:52]: I guess this change we could actually keep. The old form probably could not use dict comprehension because it had to support Python 2.6.

File: tests/test_sinks.py
Comment (vytas7) [inline tests/test_sinks.py:26]: Reimplementing `__call__` is intended here (note the change from sync to `async`).

**Review by vytas7 — CHANGES_REQUESTED** [review]
Thank you once again for this initiative to improve our code quality.

However, as discussed in the above comments, we do not want to indiscriminately refactor the following:
* Functions not making use of `self` :arrow_right: `@staticmethod`. `@staticmethod` may at times semantically make sense, but more often than not, it is not the case in our codebase. Moreover, it does not offer any speed improvement (it actually may even incur a slight penalty).
* `assert` :arrow_right: manually raising an `AssertionError` in a branch statement -- we simply do not want this.
* `elif` after `return` or `raise` :arrow_right:  `if`. While I have nothing against the pattern itself, and actually use it, we need to discuss whether it is worth the effort (e.g., how do these alternatives compare performance wise?) refactoring. I.e., I'm actually not opposed to this one, but it is better to discuss this one separately.

I know it feels like harsh feedback... :imp:  But could you roll back refactoring of the 3 above patterns, and we could take a look at what's left. As noted on the PR, there are some good changes that we could actually be willing to keep.

Needless to say, @withshubh, You are also very welcome to discuss these changes with us on [our Gitter channels](https://falconframework.org/#sectionCommunity)!

**Comment by vytas7** [comment]
@withshubh just checking, if you are planning to continue work on this PR?

And I'm sorry to be blunt (again), but an _en masse_ `autopep8` refactoring is not something we want to accept as a side effect of this PR. We are looking into adopting `black`, `autopep8` or other alternatives, but that needs to be a separate, informed decision by the maintainer team. You'll have to revert those changes from your branch in order to proceed.

**Comment by withshubh** [comment]
Hi @vytas7 :wave: 

It's okay for being blunt(again) :stuck_out_tongue: I'll fix this and update the PR. :sparkles:

**Comment by withshubh** [comment]
Hi @vytas7 

Please have a look. :eyes:

File: .deepsource.toml
Comment (vytas7) [inline .deepsource.toml:1]: Please remove this file.
We have not decided on using DeepSource yet.

File: falcon/util/structures.py
Comment (vytas7) [inline falcon/util/structures.py:64]: Please read the comment on line 33 to understand why, and revert this change.

File: falcon/vendor/mimeparse/mimeparse.py
Comment (vytas7) [inline falcon/vendor/mimeparse/mimeparse.py:101]: Let us revert these changes. Sensible they may be, we have vendored this module from https://github.com/dbtsai/python-mimeparse, and do not want to make any modifications unless absolutely necessary.

File: falcon/vendor/mimeparse/mimeparse.py
Comment (vytas7) [inline falcon/vendor/mimeparse/mimeparse.py:184]: Let us revert these changes. Sensible they may be, we have vendored this module from https://github.com/dbtsai/python-mimeparse, and do not want to make any modifications unless absolutely necessary.

**Review by vytas7 — CHANGES_REQUESTED** [review]
Please see inline comments.

**Comment by withshubh** [comment]
Hi @vytas7 

Now this PR contains just one fix:

- remove unnecessary generator

I'd like to know what are your thoughts regarding integrating the DeepSource.

**Review by vytas7 — CHANGES_REQUESTED** [review]
Let us please address PEP8 violations before merging this PR:
```
./falcon/util/uri.py:53:21: E127 continuation line over-indented for visual indent
./falcon/util/uri.py:54:21: E127 continuation line over-indented for visual indent
```

**Comment by withshubh** [comment]
> Let us please address PEP8 violations before merging this PR:
> 
> ```
> ./falcon/util/uri.py:53:21: E127 continuation line over-indented for visual indent
> ./falcon/util/uri.py:54:21: E127 continuation line over-indented for visual indent
> ```

Done @vytas7 :sparkles:

**Review by vytas7 — APPROVED** [review]
Thanks for this!

Regarding DeepSource, my impression is that it is likely a very useful tool for a business project, or even a new open source library if DeepSource is agreed upon in advance.
However, for a highly specialized framework Falcon is, as you may have noticed, many "quality issues" exist there for a reason ("vendored" code, explicit test scenarios, deprecation strategies etc).
Moreover, another drawback is that DeepSource is not open, so we cannot easily integrate it into our offline development routines.

**Pull request merged**
